### ISSUE TITLE
Travis CI build is broken for most macOS versions

### ISSUE THREAD

**Issue by dgelessus** [beeware/rubicon-objc#141]
Example: https://travis-ci.org/beeware/rubicon-objc/builds/630497088

For all Travis environments except `xcode10.2` (the most recent one that our CI configuration includes),  the installation of Python 3.5 (using `pyenv`) fails with this message:

```
Last 10 log lines:
    import pip
  File "/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/tmprxyor7dg/pip-9.0.1-py2.py3-none-any.whl/pip/__init__.py", line 26, in <module>
  File "/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/tmprxyor7dg/pip-9.0.1-py2.py3-none-any.whl/pip/utils/__init__.py", line 27, in <module>
  File "/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/tmprxyor7dg/pip-9.0.1-py2.py3-none-any.whl/pip/_vendor/pkg_resources/__init__.py", line 35, in <module>
  File "/private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/python-build.20191231013638.23409/Python-3.5.6/Lib/plistlib.py", line 65, in <module>
    from xml.parsers.expat import ParserCreate
  File "/private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/python-build.20191231013638.23409/Python-3.5.6/Lib/xml/parsers/expat.py", line 4, in <module>
    from pyexpat import *
ImportError: No module named 'pyexpat'
make: *** [install] Error 1
```

I haven't looked into this in detail yet. Clearly the `pyexpat` module is missing, which likely means that it was not built during the CPython build. This usually happens when a required native library is not installed or cannot be found during build. I don't know what needs to be done to fix this - best case we only need to install the missing library with `brew`, but it might be more complicated than that (especially because all of these build environments were previously working).

I also don't know if this problem is specific to Python 3.5 or if it also affects later versions. Right now it's impossible to tell, because Python 3.5 is installed first, and the installation failure stops the entire build.

**Comment by freakboy3742** [comment]
/me grumbles...

This might be what pushes me over the edge to configure Github Actions for CI. Actions has much nicer handling of Python versions, and the UX is a lot nicer (because it's integrated with Github itself).

The complication is supporting old macOS versions in the build matrix - AFAICT, we can only run on the most recent macOS, with whatever historical versions of Xcode are supported (currently, back to 9.4 AFAICT). I'm trying to recall if that's a major issue, though... my recollection is that the problems we've seen have more to do with Python versioning than macOS versioning, but I could be remembering incorrectly...

**Comment by dgelessus** [comment]
> This might be what pushes me over the edge to configure Github Actions for CI.

I thought about GitHub Actions too, but AFAICT they currently only support macOS 10.15. I'm sure that will solve all our Python installation issues, but we also lose basically our entire test matrix - and for something like Rubicon that makes lots of direct calls to low-level system functions, testing on older macOSes is important.

I've looked at macOS support in other CI services a while ago, and my impression was that Travis CI's macOS support is actually fairly good in comparison, especially the range of supported macOS/Xcode versions. Unfortunately this large range of versions also makes it difficult to install software reliably in all environments...

> The complication is supporting old macOS versions in the build matrix - AFAICT, we can only run on the most recent macOS, with whatever historical versions of Xcode are supported (currently, back to 9.4 AFAICT).

According to the comments in the .travis.yml, the macOS version used by the build is determined by the selected Xcode version:

https://github.com/beeware/rubicon-objc/blob/5ca7dd39d880dcbd404306d4072aa9338959050b/.travis.yml#L23-L31

And looking at the build logs, the comments seem to still be correct - it seems that Travis hasn't retroactively updated any of the build environments that we use.

I'm guessing that something has changed either in Homebrew or pyenv, and as a result the libraries needed by `pyexpat` are no longer installed. I'm not sure how to effectively debug this though.

**Comment by freakboy3742** [comment]
Right - so my underlying question: how much do we actually care about testing across macOS versions (and old macOS versions in particular)? Are we catching (or have we historically caught) errors by testing on old macOS versions? My recollection is that *Python* version changes are a more common source of errors - ranging from forgetting about feature availability (e.g.,  a stray f-string blocking on 3.5 builds, to issues with ctypes changes. The level at which we're interacting with macOS is mostly dealing with very old and stable features of Objective C, and as a result, the macOS version changes aren't as big a concern.

There's also the question of how vigilant we need to be in supporting old versions of macOS. Officially, Sierra is currently unsupported (as of October 2019) - should we bother tying ourselves in knots to support a version of macOS that that Apple *itself* doesn't support?

**Comment by dgelessus** [comment]
Hm, you might be right - after looking through the commit history a little, it seems we don't run into macOS-version-specific problems very often. Those problems probably affect downstream users more than Rubicon itself, because Apple changes the frameworks more often than the low-level ObjC runtime. For Rubicon itself, different architectures are a bigger issue, but I'm not aware of any CI service that offers free testing on iOS.

Regarding what old macOS versions to support: Apple's macOS support timelines are quite short and in my experience don't match what users are actually running. According to [some] random (https://www.statista.com/statistics/944559/worldwide-macos-version-market-share/) [statistics](https://gs.statcounter.com/macos-version-market-share/desktop/worldwide) I found, roughly 18 % of users run macOS versions that are unsupported (at the time the stats are dated). Even OS X 10.10 (Yosemite) alone (which was released in 2014 and unsupported since 2017), has a usage share of about 4.5 %, which is [comparable to Windows 8.1](https://gs.statcounter.com/windows-version-market-share/desktop/worldwide/) (which is still supported until 2023).

My usual approach to old version support is "whatever is possible with reasonable amounts of work". Since we're not running into macOS version compatibility issues very often, the amount of work needed doesn't seem to be very high :) I remember a while back I tried running Rubicon on OS X 10.7, which is quite old, and I only had to make a single change for compatibility (see #61).

Of course if the old CI environments are difficult/impossible to run reliably, I'm fine with dropping those from the matrix. I'm all for supporting old macOS versions, but it shouldn't block work on the rest of the project unnecessarily.
